graph/memstore/quadstore.go

```diff
@@ -198,24 +208,36 @@ func (qs *QuadStore) resolveVal(v quad.Value, add bool) (int64, bool) {
 		n = n[len(internalBNodePrefix):]
 		id, err := strconv.ParseInt(string(n), 10, 64)
 		if err == nil && id != 0 {
+			qs.prw.RLock()
 			if p, ok := qs.prim[id]; ok || !add {
+				qs.prw.RUnlock()
 				if add {
 					p.refs++
 				}
 				return id, ok
 			}
+			qs.prw.RUnlock()
 			qs.appendPrimitive(&Primitive{ID: id, refs: 1})
 			return id, true
 		}
 	}
 	vs := v.String()
+	qs.vrw.RLock()
 	if id, exists := qs.vals[vs]; exists || !add {
```

Comment:
If the mutexes are that granular, you would probably want to wrap accesses to `qs.vals` (and other things) into separate functions, so you don't have to put Lock/Unlock everywhere.